src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -957,6 +957,82 @@ static double validateAndParseReal(String columnName, Object input, long insertR
         columnName, input.getClass(), "REAL", new String[] {"Number", "String"}, insertRowIndex);
   }
 
+  /**
+   * Validates and parses input Iceberg INT column. Allowed Java types:
+   *
+   * <ul>
+   *   <li>Number
+   *   <li>String
+   * </ul>
+   *
+   * @param columnName Column name, used in validation error messages
+   * @param input Object to validate and parse
+   * @param insertRowIndex Row index for error reporting
+   * @return Parsed integer
+   */
+  static int validateAndParseIcebergInt(String columnName, Object input, long insertRowIndex) {
+    if (input instanceof Number) {
+      double value = ((Number) input).doubleValue();
+      long longValue = Math.round(value);
+      if (longValue > Integer.MAX_VALUE || longValue < Integer.MIN_VALUE || Double.isNaN(value)) {
+        throw new SFException(
+            ErrorCode.INVALID_VALUE_ROW,
+            String.format(
+                "Number out of representable inclusive range of integers between %d and %d,"
+                    + " rowIndex:%d",
+                Integer.MIN_VALUE, Integer.MAX_VALUE, insertRowIndex));
+      }
+      return (int) longValue;
+    } else if (input instanceof String) {
+      try {
+        return Integer.parseInt(((String) input).trim());
+      } catch (NumberFormatException e) {
+        throw valueFormatNotAllowedException(
+            columnName, "INT", "Not a valid integer", insertRowIndex);
+      }
+    }
+    throw typeNotAllowedException(
+        columnName, input.getClass(), "INT", new String[] {"Number", "String"}, insertRowIndex);
+  }
+
+  /**
+   * Validates and parses input Iceberg LONG column. Allowed Java types:
+   *
+   * <ul>
+   *   <li>Number
+   *   <li>String
+   * </ul>
+   *
+   * @param columnName Column name, used in validation error messages
+   * @param input Object to validate and parse
+   * @param insertRowIndex Row index for error reporting
+   * @return Parsed long
+   */
+  static long validateAndParseIcebergLong(String columnName, Object input, long insertRowIndex) {
+    if (input instanceof Number) {
```

Comment:
Should we return right away for integer types without going via double?

Reply:
Fixed. Used the logic from `validateAndParseBigDecimal`.